digest/README.md

```diff
@@ -0,0 +1,254 @@
+`my_ingress_metadata_t` will hold the port on where the packet arrived and `send_digest`. `send_digest` signals to the deparser to send a digest message to user space. This digest message will hold the source mac address and the port where the packet arrived:
+
+```
+struct mac_learn_digest_t {
+    @tc_type("macaddr") bit<48> srcAddr;
+    @tc_type("dev") PortId_t ingress_port;
+    DROP_REASON drop_reason;
+};
+```
+
+## Setup Requirements
+
+Make sure that the p4node basic container has already been created at this point(as per instructions found in [p4node](https://github.com/p4tc-dev/p4tc-examples-pub.git)). To run the sample described setup here requires 4 terminals, three terminals inside the container and one on the VM side to generate traffic.
+
+### Terminal 1 (observation of tc commands on p4node).
+
+Enter the container p4node:
+
+`sudo ip netns exec p4node /bin/bash`
+
+setup path for TC binary
+
+`TC="/usr/sbin/tc"`
+
+setup the path to where the json introspection file can be found..
+
+```
+cd /home/vagrant/p4tc-examples-pub/digest/generated
+export INTROSPECTION=.
+```
+run TC monitor:
+`$TC mon p4 events`
+
+### Terminal 2 (observation of p4 digest events on p4node).
+
+First enter the container and make sure you have the introspection path setup
+
+Enter the container p4node:
+
+`sudo ip netns exec p4node /bin/bash`
+
+setup path for TC binary
+
+`TC="/usr/sbin/tc"`
+
+setup the path to where the json introspection file can be found..
+
+```
+cd /home/vagrant/p4tc-examples-pub/digest/generated
+export INTROSPECTION=.
+```
+
+run TC monitor for digest events:
+`$TC mon p4 digest`
+
+### Terminal 3 (observes incoming traffic into p4node)
+
+First enter the container and make sure you have the introspection path setup
+
+`sudo ip netns exec p4node /bin/bash`
+
+Now let's listen to traffic on port0
+
+```
+DEV=port0
+tcpdump -n -i $DEV -e
+```
+
+### Terminal 4 (to instantiate and runtime control the program)
+
+We will run commands to first load the prog and then do required runtime setup.
+
+First enter the container
+
+```
+sudo ip netns exec p4node /bin/bash
+cd /home/vagrant/p4tc-examples-pub/digest
+```
+
+```
+modprobe ext_Digest
+```
+
+Compile the parser and control blocks programs if you have not already
+
+`make`
+
+Make sure you have the introspection path setup and load the *digest* program
+
+```
+cd generated
+export INTROSPECTION=.
+TC="/usr/sbin/tc"
+./digest.template
+```
+
+now instantiate the prog
+
+```
+$TC filter add block 21 ingress protocol all prio 10 p4 pname digest \
+action bpf obj digest_parser.o section p4tc/parse \
+action bpf obj digest_control_blocks.o section p4tc/main
+```
+
+### Terminal 5 (on the VM side)
+
+Try sending a message of packets which will be dropped by the parser (observe tcpdump on terminal 3)..
+
+`ping -I p4port0 10.0.1.2 -c 1`
+
+Let's check some stats, below shows 3 packets dropped by the parser on <u>terminal 3</u>:
+
```

Comment:
Shouldnt you point the user to terminal 2 at this point? And what they should expect to see?
And what was the packet dropped by the parser.
You are literally cutting and pasting other examples i had without thinking through what useful information is needed for the user..

Reply:
That's why I transformed it into a draft, because I needed to fix it given the new semantic